# eDNA conversion of practice data - draft 4

**Resources:**
- https://docs.gbif-uat.org/publishing-dna-derived-data/1.0/en/#mapping-metabarcoding-edna-and-barcoding-data

In [1]:
## Imports

import pandas as pd
import numpy as np
import random

from datetime import datetime # for handling dates
import pytz # for handling time zones

In [2]:
## Ensure my general functions for the MPA data integration project can be imported, and import them

import sys
sys.path.insert(0, '/Users/dianalg/PycharmProjects/PythonScripts/eDNA')

import WoRMS # functions for querying WoRMS REST API

## Load data

In [3]:
## Plate data

plate = pd.read_csv('Plate_S_ASV_OBIS_data.csv')
print(plate.shape)
plate.head()

(280440, 11)


ASV            FilterID  \
0  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  05114c01_12_edna_1   
1  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  05114c01_12_edna_2   
2  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  05114c01_12_edna_3   
3  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  11216c01_12_edna_1   
4  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  11216c01_12_edna_2   

            Sequence_ID  Reads    Kingdom      Phylum        Class      Order  \
0  05114c01_12_edna_1_S  14825  Eukaryota  Arthropoda  Hexanauplia  Calanoida   
1  05114c01_12_edna_2_S  16094  Eukaryota  Arthropoda  Hexanauplia  Calanoida   
2  05114c01_12_edna_3_S  22459  Eukaryota  Arthropoda  Hexanauplia  Calanoida   
3  11216c01_12_edna_1_S  19312  Eukaryota  Arthropoda  Hexanauplia  Calanoida   
4  11216c01_12_edna_2_S  16491  Eukaryota  Arthropoda  Hexanauplia  Calanoida   

          Family        Genus     Species  
0  Paracalanidae  Paracalanus  unassigned  
1  Paracalanidae  Paracalanus  unassigned  
2  Paracalanidae  Paracalanus  unassigned  
3  Paracalanidae  Paracalanus  unassigned  
4  Paracalanidae  Paracalanus  unassigned

**FilterID** = SAMPLING_cruise (cruise where the sample was collected) + 
    SAMPLING_station_number (in this case c + a cast number because these data were collected via CTD cast) + _ +
    SAMPLING_bottle (bottle number, if multiple samples were taken from a single bottle, there may be replicates indicated by a, b, c, etc.) + _ +
    edna (filter indicator. edna means that the standard filter for the standard edna pipeline, i.e. a standard PBDF filter, was used. other options include hplc) + _ +
    1, 2 or 3 (PCR replicate)

To summarize, **FilterID** = cruise number + cast number + bottle number + filter indicator + replicate number.

**Note** that not all IDs exactly follow this formula. Example:
- JD10706C1_0m_1

**Sequence_ID** adds the plate indicator to the end of FilterID. In this case, all samples are from plate S. Plates are labeled A-Z, AA-ZZ, etc.

In [4]:
## Taxonomy table

taxa = pd.read_csv('Filtered_ASV_taxa_table_all.csv')
print(taxa.shape)
taxa.head()

(4711, 79)


ASV    Kingdom      Phylum        Class          Order          Family  \
0  ASV_1  Eukaryota  Arthropoda  Hexanauplia      Calanoida   Paracalanidae   
1  ASV_2  Eukaryota  Arthropoda  Hexanauplia      Calanoida       Calanidae   
2  ASV_3  Eukaryota     unknown  Dinophyceae  Gymnodiniales  Gymnodiniaceae   
3  ASV_4  Eukaryota     unknown  Dinophyceae  Gymnodiniales  Gymnodiniaceae   
4  ASV_5  Eukaryota     unknown  Dinophyceae     unassigned      unassigned   

          Genus             Species  14213c01_12_edna_1  14213c01_12_edna_2  \
0   Paracalanus          unassigned                   9                   4   
1    unassigned          unassigned                   4                   7   
2      Akashiwo  Akashiwo sanguinea                   2                   3   
3  Cochlodinium          unassigned                   2                   3   
4    unassigned          unassigned                 148                 147   

   ...  JD33306C1_0m_3  pcrblank2  EB_20161121  EB_20161228  EB_20170117  \
0  ...             956          0            0         1312            1   
1  ...             358          0            0          535            0   
2  ...             795          0            0            0           35   
3  ...           82065          0            0            0          836   
4  ...           14688          0            0           36          401   

   pcrblank3  CB_CANON160925_1  CB_CANON160925_2  CB_CANON160925_3  ArtComm2  
0          0               197                 3               210        44  
1          0               118               162               139      5976  
2          0                 1                 0                 1         8  
3          0                 1                 0                 0         7  
4          0                 0                 4                11         7  

[5 rows x 79 columns]

This shows the number of reads for each ASV detected for each FilterID (columns), plus the matched taxa information for each ASV, if available. 

Some columns are associated with controls:
- **CB** = collection blank which was taken on the ship. Water that should be "clean" was passed through the filter, so there should be no reads in this sample, although it tends to be the "dirtiest" control. It checks for contamination during filtration, I think.
- **EB** = extraction blank which was taken in the lab but contains no DNA (i.e. a negative control), so there should be no reads in this sample. It checks if contamination occurred during DNA extraction (lab work pre-PCR).
- **pcrblank** = A sample that went through PCR but contained no input DNA (i.e. a negative control), so there should be no reads in this sample. It checks if contamination occurred during PCR.
- **ArtComm** = artificial community that went through PCR and contained DNA from species that should not be in the study system (i.e. a positive control), so you know what results you expect.

To see these columns, use:
```python
for col in taxa.columns:
    if col not in plate['FilterID'].unique():
        print(col)
```

Additionally, taxonomic columns include the following designations:
- **unknown** = GenBank couldn't give a scientifically-agreed-upon name for a given taxonomic level. I.e., either the name doesn't exist, or there isn't enough scientific consensus to give a name.
- **no_hit** = BLAST did not find any hits for the ASV.
- **unassigned** = The ASV got BLAST hits, but the post-processing program Megan6 didn't assign the ASV to any taxonomic group.
- **g_** or **s_** = After Megan there's an additional filtering step where Genus or Species designations are only left if they were assigned with a certain, high level of confidence or higher. If this threshold is not met, a g_ or s_ is given.

In [5]:
## Plate metadata

plate_meta = pd.read_csv('Plate_S_meta_OBIS_data_071521.csv')
print(plate_meta.shape)
plate_meta.head()

(60, 67)


sample_name library  tag_sequence primer_sequence_forward  \
0  14213c01_12_eDNA_1      S1  ACGAGACTGATT        GTACACACCGCCCGTC   
1  14213c01_12_eDNA_2      S2  GAATACCAAGTC        GTACACACCGCCCGTC   
2  14213c01_12_eDNA_3      S3  CGAGGGAAAGTC        GTACACACCGCCCGTC   
3  22013c01_12_eDNA_1      S4  GAACACTTTGGA        GTACACACCGCCCGTC   
4  22013c01_12_eDNA_2      S5  ACTCACAGGAAT        GTACACACCGCCCGTC   

    primer_sequence_reverse                                          R1  \
0  TGATCCTTCTGCAGGTTCACCTAC  14213c01_12_edna_1_S1_L001_R1_001.fastq.gz   
1  TGATCCTTCTGCAGGTTCACCTAC  14213c01_12_edna_2_S2_L001_R1_001.fastq.gz   
2  TGATCCTTCTGCAGGTTCACCTAC  14213c01_12_edna_3_S3_L001_R1_001.fastq.gz   
3  TGATCCTTCTGCAGGTTCACCTAC  22013c01_12_edna_1_S4_L001_R1_001.fastq.gz   
4  TGATCCTTCTGCAGGTTCACCTAC  22013c01_12_edna_2_S5_L001_R1_001.fastq.gz   

                                           R2 PlateID    sample_type  \
0  14213c01_12_edna_1_S1_L001_R2_001.fastq.gz       S  environmental   
1  14213c01_12_edna_2_S2_L001_R2_001.fastq.gz       S  environmental   
2  14213c01_12_edna_3_S3_L001_R2_001.fastq.gz       S  environmental   
3  22013c01_12_edna_1_S4_L001_R2_001.fastq.gz       S  environmental   
4  22013c01_12_edna_2_S5_L001_R2_001.fastq.gz       S  environmental   

  target_gene  ...  pcr_primer_name_forward pcr_primer_name_reverse  \
0         18S  ...                    1391f                   EukBr   
1         18S  ...                    1391f                   EukBr   
2         18S  ...                    1391f                   EukBr   
3         18S  ...                    1391f                   EukBr   
4         18S  ...                    1391f                   EukBr   

                                pcr_primer_reference            seq_meth  \
0  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...  NGS Illumina Miseq   
1  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...  NGS Illumina Miseq   
2  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...  NGS Illumina Miseq   
3  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...  NGS Illumina Miseq   
4  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...  NGS Illumina Miseq   

   sequencing_facility                 seqID  \
0             Stanford  14213c01_12_edna_1_S   
1             Stanford  14213c01_12_edna_2_S   
2             Stanford  14213c01_12_edna_3_S   
3             Stanford  22013c01_12_edna_1_S   
4             Stanford  22013c01_12_edna_2_S   

                      identificationRemarks  \
0  Genbank nr Release 221 September 20 2017   
1  Genbank nr Release 221 September 20 2017   
2  Genbank nr Release 221 September 20 2017   
3  Genbank nr Release 221 September 20 2017   
4  Genbank nr Release 221 September 20 2017   

                            identificationReferences          FilterID  \
0  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...  14213c01_12_edna   
1  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...  14213c01_12_edna   
2  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...  14213c01_12_edna   
3  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...  22013c01_12_edna   
4  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...  22013c01_12_edna   

  associatedSequences  
0                 NaN  
1                 NaN  
2                 NaN  
3                 NaN  
4                 NaN  

[5 rows x 67 columns]

## Conversion

In [30]:
## eventID

occ = pd.DataFrame({'eventID':plate['Sequence_ID']})
print(occ.shape)
occ.head()

(280440, 1)


eventID
0  05114c01_12_edna_1_S
1  05114c01_12_edna_2_S
2  05114c01_12_edna_3_S
3  11216c01_12_edna_1_S
4  11216c01_12_edna_2_S

In [31]:
## Merge with plate_meta to obtain eventDate, decimalLatitude, decimalLongitude, and other columns that can be added directly from metadata

metadata_cols = [
    'seqID',
    'eventDate', 
    'decimalLatitude', 
    'decimalLongitude',
    'env_broad_scale',
    'env_local_scale',
    'env_medium',
    'target_gene',
    'primer_sequence_forward',
    'primer_sequence_reverse',
    'pcr_primer_name_forward',
    'pcr_primer_name_reverse',
    'pcr_primer_reference',
    'sop',
    'seq_meth',
    'samp_vol_we_dna_ext',
    'nucl_acid_ext', 
    'nucl_acid_amp',
]

dwc_cols = metadata_cols.copy()
dwc_cols[0] = 'eventID'

occ = occ.merge(plate_meta[metadata_cols], how='left', left_on='eventID', right_on='seqID')
occ.drop(columns='seqID', inplace=True)
occ.columns = dwc_cols
occ.head()

eventID      eventDate  decimalLatitude  decimalLongitude  \
0  05114c01_12_edna_1_S  2/20/14 15:33          36.7958          -121.848   
1  05114c01_12_edna_2_S  2/20/14 15:33          36.7958          -121.848   
2  05114c01_12_edna_3_S  2/20/14 15:33          36.7958          -121.848   
3  11216c01_12_edna_1_S  4/21/16 14:39          36.7962          -121.846   
4  11216c01_12_edna_2_S  4/21/16 14:39          36.7962          -121.846   

                 env_broad_scale                env_local_scale  \
0  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   
1  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   
2  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   
3  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   
4  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   

                                      env_medium target_gene  \
0  waterborne particulate matter (ENVO:01000436)         18S   
1  waterborne particulate matter (ENVO:01000436)         18S   
2  waterborne particulate matter (ENVO:01000436)         18S   
3  waterborne particulate matter (ENVO:01000436)         18S   
4  waterborne particulate matter (ENVO:01000436)         18S   

  primer_sequence_forward   primer_sequence_reverse pcr_primer_name_forward  \
0        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   
1        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   
2        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   
3        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   
4        GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   

  pcr_primer_name_reverse                               pcr_primer_reference  \
0                   EukBr  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...   
1                   EukBr  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...   
2                   EukBr  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...   
3                   EukBr  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...   
4                   EukBr  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...   

                                                 sop            seq_meth  \
0  dx.doi.org/10.17504/protocols.io.xjufknw|dx.do...  NGS Illumina Miseq   
1  dx.doi.org/10.17504/protocols.io.xjufknw|dx.do...  NGS Illumina Miseq   
2  dx.doi.org/10.17504/protocols.io.xjufknw|dx.do...  NGS Illumina Miseq   
3  dx.doi.org/10.17504/protocols.io.xjufknw|dx.do...  NGS Illumina Miseq   
4  dx.doi.org/10.17504/protocols.io.xjufknw|dx.do...  NGS Illumina Miseq   

  samp_vol_we_dna_ext                             nucl_acid_ext  \
0               100ml  dx.doi.org/10.17504/protocols.io.xjufknw   
1               100ml  dx.doi.org/10.17504/protocols.io.xjufknw   
2               100ml  dx.doi.org/10.17504/protocols.io.xjufknw   
3              1000ml  dx.doi.org/10.17504/protocols.io.xjufknw   
4              1000ml  dx.doi.org/10.17504/protocols.io.xjufknw   

                              nucl_acid_amp  
0  dx.doi.org/10.17504/protocols.io.n2vdge6  
1  dx.doi.org/10.17504/protocols.io.n2vdge6  
2  dx.doi.org/10.17504/protocols.io.n2vdge6  
3  dx.doi.org/10.17504/protocols.io.n2vdge6  
4  dx.doi.org/10.17504/protocols.io.n2vdge6

In [32]:
## Clean sop

occ['sop'] = occ['sop'].str.replace('|', ' | ', regex=False)
occ['sop'].iloc[0]

'dx.doi.org/10.17504/protocols.io.xjufknw | dx.doi.org/10.17504/protocols.io.n2vdge6 | https://github.com/MBARI-BOG/BOG-Banzai-Dada2-Pipeline'

In [33]:
## Change column names as needed

occ = occ.rename(columns = {'primer_sequence_forward':'pcr_primer_forward',
                            'primer_sequence_reverse':'pcr_primer_reverse'})
occ.head()

eventID      eventDate  decimalLatitude  decimalLongitude  \
0  05114c01_12_edna_1_S  2/20/14 15:33          36.7958          -121.848   
1  05114c01_12_edna_2_S  2/20/14 15:33          36.7958          -121.848   
2  05114c01_12_edna_3_S  2/20/14 15:33          36.7958          -121.848   
3  11216c01_12_edna_1_S  4/21/16 14:39          36.7962          -121.846   
4  11216c01_12_edna_2_S  4/21/16 14:39          36.7962          -121.846   

                 env_broad_scale                env_local_scale  \
0  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   
1  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   
2  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   
3  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   
4  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   

                                      env_medium target_gene  \
0  waterborne particulate matter (ENVO:01000436)         18S   
1  waterborne particulate matter (ENVO:01000436)         18S   
2  waterborne particulate matter (ENVO:01000436)         18S   
3  waterborne particulate matter (ENVO:01000436)         18S   
4  waterborne particulate matter (ENVO:01000436)         18S   

  pcr_primer_forward        pcr_primer_reverse pcr_primer_name_forward  \
0   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   
1   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   
2   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   
3   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   
4   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC                   1391f   

  pcr_primer_name_reverse                               pcr_primer_reference  \
0                   EukBr  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...   
1                   EukBr  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...   
2                   EukBr  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...   
3                   EukBr  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...   
4                   EukBr  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...   

                                                 sop            seq_meth  \
0  dx.doi.org/10.17504/protocols.io.xjufknw | dx....  NGS Illumina Miseq   
1  dx.doi.org/10.17504/protocols.io.xjufknw | dx....  NGS Illumina Miseq   
2  dx.doi.org/10.17504/protocols.io.xjufknw | dx....  NGS Illumina Miseq   
3  dx.doi.org/10.17504/protocols.io.xjufknw | dx....  NGS Illumina Miseq   
4  dx.doi.org/10.17504/protocols.io.xjufknw | dx....  NGS Illumina Miseq   

  samp_vol_we_dna_ext                             nucl_acid_ext  \
0               100ml  dx.doi.org/10.17504/protocols.io.xjufknw   
1               100ml  dx.doi.org/10.17504/protocols.io.xjufknw   
2               100ml  dx.doi.org/10.17504/protocols.io.xjufknw   
3              1000ml  dx.doi.org/10.17504/protocols.io.xjufknw   
4              1000ml  dx.doi.org/10.17504/protocols.io.xjufknw   

                              nucl_acid_amp  
0  dx.doi.org/10.17504/protocols.io.n2vdge6  
1  dx.doi.org/10.17504/protocols.io.n2vdge6  
2  dx.doi.org/10.17504/protocols.io.n2vdge6  
3  dx.doi.org/10.17504/protocols.io.n2vdge6  
4  dx.doi.org/10.17504/protocols.io.n2vdge6

In [34]:
## Add extension terms that weren't in metadata file

occ['target_subfragment'] = 'V9'
occ['lib_layout'] = 'paired'
occ['otu_seq_comp_appr'] = 'blast;version;80% identity | MEGAN6;version; bitscore:100:2%'
occ['otu_db'] = 'Genbank nr Release 221 September 20 2017'

occ.head()

eventID      eventDate  decimalLatitude  decimalLongitude  \
0  05114c01_12_edna_1_S  2/20/14 15:33          36.7958          -121.848   
1  05114c01_12_edna_2_S  2/20/14 15:33          36.7958          -121.848   
2  05114c01_12_edna_3_S  2/20/14 15:33          36.7958          -121.848   
3  11216c01_12_edna_1_S  4/21/16 14:39          36.7962          -121.846   
4  11216c01_12_edna_2_S  4/21/16 14:39          36.7962          -121.846   

                 env_broad_scale                env_local_scale  \
0  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   
1  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   
2  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   
3  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   
4  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   

                                      env_medium target_gene  \
0  waterborne particulate matter (ENVO:01000436)         18S   
1  waterborne particulate matter (ENVO:01000436)         18S   
2  waterborne particulate matter (ENVO:01000436)         18S   
3  waterborne particulate matter (ENVO:01000436)         18S   
4  waterborne particulate matter (ENVO:01000436)         18S   

  pcr_primer_forward        pcr_primer_reverse  ...  \
0   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   
1   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   
2   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   
3   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   
4   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   

                                pcr_primer_reference  \
0  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...   
1  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...   
2  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...   
3  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...   
4  Amaral-Zettler LA, McCliment EA, Ducklow HW, H...   

                                                 sop            seq_meth  \
0  dx.doi.org/10.17504/protocols.io.xjufknw | dx....  NGS Illumina Miseq   
1  dx.doi.org/10.17504/protocols.io.xjufknw | dx....  NGS Illumina Miseq   
2  dx.doi.org/10.17504/protocols.io.xjufknw | dx....  NGS Illumina Miseq   
3  dx.doi.org/10.17504/protocols.io.xjufknw | dx....  NGS Illumina Miseq   
4  dx.doi.org/10.17504/protocols.io.xjufknw | dx....  NGS Illumina Miseq   

  samp_vol_we_dna_ext                             nucl_acid_ext  \
0               100ml  dx.doi.org/10.17504/protocols.io.xjufknw   
1               100ml  dx.doi.org/10.17504/protocols.io.xjufknw   
2               100ml  dx.doi.org/10.17504/protocols.io.xjufknw   
3              1000ml  dx.doi.org/10.17504/protocols.io.xjufknw   
4              1000ml  dx.doi.org/10.17504/protocols.io.xjufknw   

                              nucl_acid_amp target_subfragment lib_layout  \
0  dx.doi.org/10.17504/protocols.io.n2vdge6                 V9     paired   
1  dx.doi.org/10.17504/protocols.io.n2vdge6                 V9     paired   
2  dx.doi.org/10.17504/protocols.io.n2vdge6                 V9     paired   
3  dx.doi.org/10.17504/protocols.io.n2vdge6                 V9     paired   
4  dx.doi.org/10.17504/protocols.io.n2vdge6                 V9     paired   

                                   otu_seq_comp_appr  \
0  blast;version;80% identity | MEGAN6;version; b...   
1  blast;version;80% identity | MEGAN6;version; b...   
2  blast;version;80% identity | MEGAN6;version; b...   
3  blast;version;80% identity | MEGAN6;version; b...   
4  blast;version;80% identity | MEGAN6;version; b...   

                                     otu_db  
0  Genbank nr Release 221 September 20 2017  
1  Genbank nr Release 221 September 20 2017  
2  Genbank nr Release 221 September 20 2017  
3  Genbank nr Release 221 September 20 2017  
4  Genbank nr Release 221 September 20 2017  

[5 rows x 22 columns]

**NOTE** that I need to check on:
- otu_seq_comp_appr - Katie needs to tell me what the correct value is
- samp_vol_we_dna_ext - check what this means with Saara and ask Katie how it's being used

In [35]:
## occurrenceID

occ['occurrenceID'] = plate.groupby('Sequence_ID')['ASV'].cumcount()+1
occ['occurrenceID'] = occ['eventID'] + '_occ' + occ['occurrenceID'].astype(str)
occ.head()

eventID      eventDate  decimalLatitude  decimalLongitude  \
0  05114c01_12_edna_1_S  2/20/14 15:33          36.7958          -121.848   
1  05114c01_12_edna_2_S  2/20/14 15:33          36.7958          -121.848   
2  05114c01_12_edna_3_S  2/20/14 15:33          36.7958          -121.848   
3  11216c01_12_edna_1_S  4/21/16 14:39          36.7962          -121.846   
4  11216c01_12_edna_2_S  4/21/16 14:39          36.7962          -121.846   

                 env_broad_scale                env_local_scale  \
0  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   
1  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   
2  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   
3  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   
4  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   

                                      env_medium target_gene  \
0  waterborne particulate matter (ENVO:01000436)         18S   
1  waterborne particulate matter (ENVO:01000436)         18S   
2  waterborne particulate matter (ENVO:01000436)         18S   
3  waterborne particulate matter (ENVO:01000436)         18S   
4  waterborne particulate matter (ENVO:01000436)         18S   

  pcr_primer_forward        pcr_primer_reverse  ...  \
0   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   
1   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   
2   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   
3   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   
4   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   

                                                 sop            seq_meth  \
0  dx.doi.org/10.17504/protocols.io.xjufknw | dx....  NGS Illumina Miseq   
1  dx.doi.org/10.17504/protocols.io.xjufknw | dx....  NGS Illumina Miseq   
2  dx.doi.org/10.17504/protocols.io.xjufknw | dx....  NGS Illumina Miseq   
3  dx.doi.org/10.17504/protocols.io.xjufknw | dx....  NGS Illumina Miseq   
4  dx.doi.org/10.17504/protocols.io.xjufknw | dx....  NGS Illumina Miseq   

  samp_vol_we_dna_ext                             nucl_acid_ext  \
0               100ml  dx.doi.org/10.17504/protocols.io.xjufknw   
1               100ml  dx.doi.org/10.17504/protocols.io.xjufknw   
2               100ml  dx.doi.org/10.17504/protocols.io.xjufknw   
3              1000ml  dx.doi.org/10.17504/protocols.io.xjufknw   
4              1000ml  dx.doi.org/10.17504/protocols.io.xjufknw   

                              nucl_acid_amp target_subfragment lib_layout  \
0  dx.doi.org/10.17504/protocols.io.n2vdge6                 V9     paired   
1  dx.doi.org/10.17504/protocols.io.n2vdge6                 V9     paired   
2  dx.doi.org/10.17504/protocols.io.n2vdge6                 V9     paired   
3  dx.doi.org/10.17504/protocols.io.n2vdge6                 V9     paired   
4  dx.doi.org/10.17504/protocols.io.n2vdge6                 V9     paired   

                                   otu_seq_comp_appr  \
0  blast;version;80% identity | MEGAN6;version; b...   
1  blast;version;80% identity | MEGAN6;version; b...   
2  blast;version;80% identity | MEGAN6;version; b...   
3  blast;version;80% identity | MEGAN6;version; b...   
4  blast;version;80% identity | MEGAN6;version; b...   

                                     otu_db               occurrenceID  
0  Genbank nr Release 221 September 20 2017  05114c01_12_edna_1_S_occ1  
1  Genbank nr Release 221 September 20 2017  05114c01_12_edna_2_S_occ1  
2  Genbank nr Release 221 September 20 2017  05114c01_12_edna_3_S_occ1  
3  Genbank nr Release 221 September 20 2017  11216c01_12_edna_1_S_occ1  
4  Genbank nr Release 221 September 20 2017  11216c01_12_edna_2_S_occ1  

[5 rows x 23 columns]

In [36]:
## Add DNA_sequence

occ['DNA_sequence'] = plate['ASV']
occ.head()

eventID      eventDate  decimalLatitude  decimalLongitude  \
0  05114c01_12_edna_1_S  2/20/14 15:33          36.7958          -121.848   
1  05114c01_12_edna_2_S  2/20/14 15:33          36.7958          -121.848   
2  05114c01_12_edna_3_S  2/20/14 15:33          36.7958          -121.848   
3  11216c01_12_edna_1_S  4/21/16 14:39          36.7962          -121.846   
4  11216c01_12_edna_2_S  4/21/16 14:39          36.7962          -121.846   

                 env_broad_scale                env_local_scale  \
0  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   
1  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   
2  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   
3  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   
4  marine biome  (ENVO:00000447)  coastal water (ENVO:00001250)   

                                      env_medium target_gene  \
0  waterborne particulate matter (ENVO:01000436)         18S   
1  waterborne particulate matter (ENVO:01000436)         18S   
2  waterborne particulate matter (ENVO:01000436)         18S   
3  waterborne particulate matter (ENVO:01000436)         18S   
4  waterborne particulate matter (ENVO:01000436)         18S   

  pcr_primer_forward        pcr_primer_reverse  ...            seq_meth  \
0   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  NGS Illumina Miseq   
1   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  NGS Illumina Miseq   
2   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  NGS Illumina Miseq   
3   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  NGS Illumina Miseq   
4   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  NGS Illumina Miseq   

  samp_vol_we_dna_ext                             nucl_acid_ext  \
0               100ml  dx.doi.org/10.17504/protocols.io.xjufknw   
1               100ml  dx.doi.org/10.17504/protocols.io.xjufknw   
2               100ml  dx.doi.org/10.17504/protocols.io.xjufknw   
3              1000ml  dx.doi.org/10.17504/protocols.io.xjufknw   
4              1000ml  dx.doi.org/10.17504/protocols.io.xjufknw   

                              nucl_acid_amp target_subfragment lib_layout  \
0  dx.doi.org/10.17504/protocols.io.n2vdge6                 V9     paired   
1  dx.doi.org/10.17504/protocols.io.n2vdge6                 V9     paired   
2  dx.doi.org/10.17504/protocols.io.n2vdge6                 V9     paired   
3  dx.doi.org/10.17504/protocols.io.n2vdge6                 V9     paired   
4  dx.doi.org/10.17504/protocols.io.n2vdge6                 V9     paired   

                                   otu_seq_comp_appr  \
0  blast;version;80% identity | MEGAN6;version; b...   
1  blast;version;80% identity | MEGAN6;version; b...   
2  blast;version;80% identity | MEGAN6;version; b...   
3  blast;version;80% identity | MEGAN6;version; b...   
4  blast;version;80% identity | MEGAN6;version; b...   

                                     otu_db               occurrenceID  \
0  Genbank nr Release 221 September 20 2017  05114c01_12_edna_1_S_occ1   
1  Genbank nr Release 221 September 20 2017  05114c01_12_edna_2_S_occ1   
2  Genbank nr Release 221 September 20 2017  05114c01_12_edna_3_S_occ1   
3  Genbank nr Release 221 September 20 2017  11216c01_12_edna_1_S_occ1   
4  Genbank nr Release 221 September 20 2017  11216c01_12_edna_2_S_occ1   

                                        DNA_sequence  
0  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  
1  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  
2  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  
3  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  
4  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...  

[5 rows x 24 columns]

In [25]:
## scientificName, taxonomic info

occ['scientificName'] = plate['Species']
occ['kingdom'] = plate['Kingdom']
occ['phylum'] = plate['Phylum']
occ['class'] = plate['Class']
occ['order'] = plate['Order']
occ['family'] = plate['Family']
occ['genus'] = plate['Genus']

occ.head()

eventID                  eventDate  decimalLatitude  \
0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   
1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   
2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   
3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   
4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   

   decimalLongitude                env_broad_scale  \
0          -121.848  marine biome  (ENVO:00000447)   
1          -121.848  marine biome  (ENVO:00000447)   
2          -121.848  marine biome  (ENVO:00000447)   
3          -121.846  marine biome  (ENVO:00000447)   
4          -121.846  marine biome  (ENVO:00000447)   

                 env_local_scale  \
0  coastal water (ENVO:00001250)   
1  coastal water (ENVO:00001250)   
2  coastal water (ENVO:00001250)   
3  coastal water (ENVO:00001250)   
4  coastal water (ENVO:00001250)   

                                      env_medium target_gene  \
0  waterborne particulate matter (ENVO:01000436)         18S   
1  waterborne particulate matter (ENVO:01000436)         18S   
2  waterborne particulate matter (ENVO:01000436)         18S   
3  waterborne particulate matter (ENVO:01000436)         18S   
4  waterborne particulate matter (ENVO:01000436)         18S   

  pcr_primer_forward        pcr_primer_reverse  ...  \
0   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   
1   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   
2   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   
3   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   
4   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   

                                                 sop  \
0  dx.doi.org/10.17504/protocols.io.xjufknw | dx....   
1  dx.doi.org/10.17504/protocols.io.xjufknw | dx....   
2  dx.doi.org/10.17504/protocols.io.xjufknw | dx....   
3  dx.doi.org/10.17504/protocols.io.xjufknw | dx....   
4  dx.doi.org/10.17504/protocols.io.xjufknw | dx....   

                occurrenceID  \
0  05114c01_12_edna_1_S_occ1   
1  05114c01_12_edna_2_S_occ1   
2  05114c01_12_edna_3_S_occ1   
3  11216c01_12_edna_1_S_occ1   
4  11216c01_12_edna_2_S_occ1   

                                        DNA_sequence scientificName  \
0  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...     unassigned   
1  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...     unassigned   
2  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...     unassigned   
3  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...     unassigned   
4  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...     unassigned   

     kingdom      phylum        class      order         family        genus  
0  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  
1  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  
2  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  
3  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  
4  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  

[5 rows x 23 columns]

In [85]:
## Replace 'unknown', 'unassigned', etc. in scientificName and taxonomy columns with NaN

cols = ['scientificName', 'kingdom', 'phylum', 'class', 'order', 'family', 'genus']
occ[cols] = occ[cols].replace({'unassigned':np.nan,
                              's_':np.nan,
                              'g_':np.nan,
                              'unknown':np.nan,
                              'no_hit':np.nan})
occ.head()

eventID                  eventDate  decimalLatitude  \
0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   
1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   
2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   
3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   
4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   

   decimalLongitude                env_broad_scale  \
0          -121.848  marine biome  (ENVO:00000447)   
1          -121.848  marine biome  (ENVO:00000447)   
2          -121.848  marine biome  (ENVO:00000447)   
3          -121.846  marine biome  (ENVO:00000447)   
4          -121.846  marine biome  (ENVO:00000447)   

                 env_local_scale  \
0  coastal water (ENVO:00001250)   
1  coastal water (ENVO:00001250)   
2  coastal water (ENVO:00001250)   
3  coastal water (ENVO:00001250)   
4  coastal water (ENVO:00001250)   

                                      env_medium target_gene  \
0  waterborne particulate matter (ENVO:01000436)         18S   
1  waterborne particulate matter (ENVO:01000436)         18S   
2  waterborne particulate matter (ENVO:01000436)         18S   
3  waterborne particulate matter (ENVO:01000436)         18S   
4  waterborne particulate matter (ENVO:01000436)         18S   

  pcr_primer_forward        pcr_primer_reverse  ...  \
0   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   
1   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   
2   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   
3   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   
4   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   

                                                 sop  \
0  dx.doi.org/10.17504/protocols.io.xjufknw | dx....   
1  dx.doi.org/10.17504/protocols.io.xjufknw | dx....   
2  dx.doi.org/10.17504/protocols.io.xjufknw | dx....   
3  dx.doi.org/10.17504/protocols.io.xjufknw | dx....   
4  dx.doi.org/10.17504/protocols.io.xjufknw | dx....   

                occurrenceID  \
0  05114c01_12_edna_1_S_occ1   
1  05114c01_12_edna_2_S_occ1   
2  05114c01_12_edna_3_S_occ1   
3  11216c01_12_edna_1_S_occ1   
4  11216c01_12_edna_2_S_occ1   

                                        DNA_sequence scientificName  \
0  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...            NaN   
1  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...            NaN   
2  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...            NaN   
3  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...            NaN   
4  GCTACTACCGATTGAACATTTTAGTGAGGTCCTCGGACTGTGAGCC...            NaN   

     kingdom      phylum        class      order         family        genus  
0  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  
1  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  
2  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  
3  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  
4  Eukaryota  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus  

[5 rows x 23 columns]

In [86]:
## Get unique species names

names = occ['scientificName'].unique()
names = names[~pd.isnull(names)] # remove NaN
print(len(names))

323


These names **do** have to be WoRMS-approved in order to be submitted to OBIS. 

There are a number of entries in the scientificName column, like "uncultured marine eukaryote", "eukaryote clone OLI11007", and "Acantharian sp. 6201", that are **not proper Latin species names**. It seems reasonable to replace these with NaN as well. **I've used a simple rule to filter these out, and it would be worth checking it as the data set expands, just to be extra careful.**

To see the entries:
```python
names = occ['scientificName'].unique()
names = names[~pd.isnull(names)] # remove NaN
for name in names:
    words_in_name = name.split(' ')
    if len(words_in_name) > 2:
        print(name)
```

In [87]:
## Replace non-Latin species names with NaN

# Get non-Latin names
non_latin_names = []
for name in names:
    words_in_name = name.split(' ')
    if len(words_in_name) > 2:
        non_latin_names.append(name)
non_latin_names_dict = {i:np.nan for i in non_latin_names}

# Add any names that didn't get caught in the simple filter
non_latin_names_dict['phototrophic eukaryote'] = np.nan
non_latin_names_dict['Candida <clade Candida/Lodderomyces clade>'] = np.nan

# Replace
occ['scientificName'].replace(non_latin_names_dict, inplace=True)

In addition, 52380 records **only give "Eukaryota" as the scientific name** (i.e. Eukaryota is in the kingdom field, and there is no more taxonomic information). These should be replaced with Biota (http://marinespecies.org/aphia.php?p=taxdetails&id=1)

In [88]:
## Replace entries where kingdom = 'Eukaryota' with the WoRMS-approved 'Biota'

occ.loc[occ['kingdom'] == 'Eukaryota', 'kingdom'] = 'Biota'

Finally **GenBank and WoRMS seem to disagree on their taxonomy**. Two examples:

1. 120 records have "Pelargonium" as the genus. [On GenBank](https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?mode=Info&id=4030&lvl=3&lin=f&keep=1&srchmode=1&unlock) the full taxonomy is given as:
    - superkingdom Eukaryota
    - kingdom Viridiplantae (subkingdom on WoRMS)
    - phylum Streptophyta (infrakingdom on WoRMS, phylum listed as Tracheophyta)
    - subphylum Streptophytina (Spermatophytina on WoRMS)
    - class Magnoliopsida (matches on WoRMS)
        - this was indicated as 'unknown' in original data, i.e. GenBank couldn't give a scientifically-agreed-upon name for a given taxonomic level. **This doesn't make sense to me.**
    - order Geraniales (matches on WoRMS)
    - family Geraniaceae ([matches on WoRMS](http://marinespecies.org/aphia.php?p=taxdetails&id=382461))
    - genus Pelargonium (does NOT match on WoRMS)


2. 600 records have family Hemistasiidae (and no further taxonomic information). [On GenBank](https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?mode=Info&id=2015512&lvl=3&lin=f&keep=1&srchmode=1&unlock), the full taxonomy is given as:
    - superkingdom Eukaryota
    - phylum Euglenozoa (matches on WoRMS)
    - order Diplonemea ([matches on WoRMS](http://marinespecies.org/aphia.php?p=taxdetails&id=582176))
    - family Hemistasiidae (does NOT match on WoRMS)

Interestingly, there are 60 records where species is Hemistasia phaeocysticola. These records also have family Hemistasiidae. The genus Hemistasia [matches on WoRMS](http://marinespecies.org/aphia.php?p=taxdetails&id=146165), but with a totally different taxonomy than [on GenBank](https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?mode=Undef&id=1503927&lvl=3&srchmode=1&keep=1&unlock):
- kingdom Protozoa
- phylum Euglenozoa
- class Kinetoplastea
- order not listed
- family not listed

**There are ongoing discussions about this issue. IOOS is working on a crosswalk between WoRMS IDs and NCBI taxonomy IDs, but it will only work for named species. See [this](https://github.com/iobis/Project-team-Genetic-Data/issues/5) GitHub issue. Either way, this is a problem that is bigger than I can solve without going case-by-case.**

**In both examples above**, the lowest taxonomic rank name given by the dataset does not match on WoRMS. I therefore need a system for searching through the higher taxonomic ranks given, finding the lowest one that will match on WoRMS, and putting that name in the scientificName column. The following few code blocks do this.

In [89]:
## Functions for finding the lowest available taxonomic rank that will match on WoRMS

def fill_lowest_taxon(df, cols):
    """ Takes the occurrence pandas data frame and fills missing values in scientificName with values from the first non-missing taxonomic rank column, listed in cols. """
    
    cols.reverse()
    
    for col in cols[:-1]:
        df['scientificName'] = df['scientificName'].combine_first(df[col])
    
    cols.reverse()
    
    return(df)

def find_not_matched(df, name_dict):
    """ Takes the occurrence pandas data frame and name_dict matching scientificName values with names on WoRMS and returns a list of names that did not match on WoRMS. """
    
    not_matched = []
    
    for name in df['scientificName'].unique():
        if name not in name_dict.keys():
            not_matched.append(name)
    
    try:
        not_matched.remove(np.nan)
    except ValueError:
        pass
            
    return(not_matched)

def replace_not_matched(df, not_matched, cols):
    """ Takes the occurrence pandas data frame and a list of scientificName values that did not match on WoRMS and replaces those values with NaN in the columns specified by cols. """
    
    df[cols] = df[cols].replace(not_matched, np.nan)
    
    return(df)  

In [90]:
## Iterate to match lowest possible taxonomic rank on WoRMS (takes ~8 minutes when starting with ~750 names)

# Note that cols (list of taxonomic column names) was defined when cleaning them above

# Initialize dictionaries
name_name_dict = {}
name_id_dict = {}
name_taxid_dict = {}
name_class_dict = {}

# Initialize not_matched
not_matched = [1]

# Iterate
while len(not_matched) > 0:
    
    # Step 1 - fill
    occ = fill_lowest_taxon(occ, cols)

    # Step 2 - get names to match
    to_match = find_not_matched(occ, name_name_dict)

    # Step 3 - match
    print('Matching {num} names on WoRMS.'.format(num = len(to_match)))
    name_id, name_name, name_taxid, name_class = WoRMS.run_get_worms_from_scientific_name(to_match, verbose_flag=False)
    name_id_dict = {**name_id_dict, **name_id}
    name_name_dict = {**name_name_dict, **name_name}
    name_taxid_dict = {**name_taxid_dict, **name_taxid}
    name_class_dict = {**name_class_dict, **name_class}
    print('Length of name_name_dict: {length}'.format(length = len(name_name_dict)))
    
    ## ---- Add in a progress bar? Also, a better/additional stopping criterion? What if not all names can be matched?
    ## ---- Also, update this code to use pyworms instead of my functions? (Could consider cloning pyworms and adding functionality)

    # Step 4 - get names that didn't match
    not_matched = find_not_matched(occ, name_name_dict)
    print('Number of names not matched: {num}'.format(num = len(not_matched)))

    # Step 5 - replace these values with NaN
    occ = replace_not_matched(occ, not_matched, cols)

Matching 756 names on WoRMS.
Length of name_name_dict: 691
Number of names not matched: 65
Matching 38 names on WoRMS.
Length of name_name_dict: 714
Number of names not matched: 15
Matching 11 names on WoRMS.
Length of name_name_dict: 719
Number of names not matched: 6
Matching 3 names on WoRMS.
Length of name_name_dict: 720
Number of names not matched: 2
Matching 1 names on WoRMS.
Length of name_name_dict: 720
Number of names not matched: 1
Matching 0 names on WoRMS.
Length of name_name_dict: 720
Number of names not matched: 0


**Note** that there are 33360 records (~12% of the data) where no taxonomic information was obtained at all (i.e. scientificName is still NaN). I've used 'Biota' for these records.

```python
occ[occ['scientificName'].isna() == True].shape
```

In [91]:
## Change scientificName to Biota in cases where all taxonomic information is missing

print(occ[occ['scientificName'].isna() == True].shape)
occ.loc[occ['scientificName'].isna() == True, 'scientificName'] = 'Biota'
occ[occ['scientificName'].isna() == True].shape

(33360, 23)


(0, 23)

Finally, during the above process, **I totally messed up the taxonomy columns in order to obtain the best possible scientificName column**. I'll fix that below.

In [92]:
## Fix taxonomy columns

# Replace with original data
occ[cols[1:]] = plate[['Kingdom', 'Phylum', 'Class', 'Order', 'Family', 'Genus']].copy()

# Replace missing data indicators in original data with empty strings ('')
occ[cols[1:]] = occ[cols[1:]].replace({
    'unassigned':'',
    's_':'',
    'g_':'',
    'unknown':'',
    'no_hit':''})

Check using:

```python
occ[occ['scientificName'] == 'Diplonemea'].head()
```

In [93]:
## Add scientific name-related columns

occ['scientificNameID'] = occ['scientificName']
occ['scientificNameID'].replace(name_id_dict, inplace=True)

occ['taxonID'] = occ['scientificName']
occ['taxonID'].replace(name_taxid_dict, inplace=True)

occ['scientificName'].replace(name_name_dict, inplace=True)

occ['nameAccordingTo'] = 'WoRMS'
occ.head()

eventID                  eventDate  decimalLatitude  \
0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   
1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   
2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   
3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   
4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   

   decimalLongitude                env_broad_scale  \
0          -121.848  marine biome  (ENVO:00000447)   
1          -121.848  marine biome  (ENVO:00000447)   
2          -121.848  marine biome  (ENVO:00000447)   
3          -121.846  marine biome  (ENVO:00000447)   
4          -121.846  marine biome  (ENVO:00000447)   

                 env_local_scale  \
0  coastal water (ENVO:00001250)   
1  coastal water (ENVO:00001250)   
2  coastal water (ENVO:00001250)   
3  coastal water (ENVO:00001250)   
4  coastal water (ENVO:00001250)   

                                      env_medium target_gene  \
0  waterborne particulate matter (ENVO:01000436)         18S   
1  waterborne particulate matter (ENVO:01000436)         18S   
2  waterborne particulate matter (ENVO:01000436)         18S   
3  waterborne particulate matter (ENVO:01000436)         18S   
4  waterborne particulate matter (ENVO:01000436)         18S   

  pcr_primer_forward        pcr_primer_reverse  ... scientificName    kingdom  \
0   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...    Paracalanus  Eukaryota   
1   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...    Paracalanus  Eukaryota   
2   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...    Paracalanus  Eukaryota   
3   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...    Paracalanus  Eukaryota   
4   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...    Paracalanus  Eukaryota   

       phylum        class      order         family        genus  \
0  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   
1  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   
2  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   
3  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   
4  Arthropoda  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   

                            scientificNameID taxonID nameAccordingTo  
0  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS  
1  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS  
2  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS  
3  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS  
4  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS  

[5 rows x 26 columns]

In [94]:
## Get set up to query NCBI taxonomy 

from Bio import Entrez
Entrez.email = 'dianalg@mbari.org'

# Get list of all databases available through this tool
record = Entrez.read(Entrez.einfo())
all_dbs = record['DbList']
all_dbs

['pubmed', 'protein', 'nuccore', 'ipg', 'nucleotide', 'structure', 'genome', 'annotinfo', 'assembly', 'bioproject', 'biosample', 'blastdbinfo', 'books', 'cdd', 'clinvar', 'gap', 'gapplus', 'grasp', 'dbvar', 'gene', 'gds', 'geoprofiles', 'homologene', 'medgen', 'mesh', 'ncbisearch', 'nlmcatalog', 'omim', 'orgtrack', 'pmc', 'popset', 'proteinclusters', 'pcassay', 'protfam', 'biosystems', 'pccompound', 'pcsubstance', 'seqannot', 'snp', 'sra', 'taxonomy', 'biocollections', 'gtr']

In [95]:
## Get NCBI taxIDs for each name in dataset ---- TAKES ~ 2 MINUTES FOR 300 RECORDS

name_ncbiid_dict = {}

for name in names:
    handle = Entrez.esearch(db='taxonomy', retmax=10, term=name)
    record = Entrez.read(handle)
    name_ncbiid_dict[name] = record['IdList'][0]
    handle.close()

**Note** that this code will throw an IndexError (IndexError: list index out of range) if a term is not found.

In [96]:
## Add NCBI taxonomy IDs under taxonConceptID

# Map indicators that say no taxonomy was assigned to empty strings
name_ncbiid_dict['unassigned'], name_ncbiid_dict['s_'], name_ncbiid_dict['no_hit'], name_ncbiid_dict['unknown'], name_ncbiid_dict['g_'] = '', '', '', '', ''

# Create column
occ['taxonConceptID']  = plate['Species'].copy()
occ['taxonConceptID'].replace(name_ncbiid_dict, inplace=True)

# Add remainder of text and clean
occ['taxonConceptID'] = 'NCBI:txid' + occ['taxonConceptID']
occ['taxonConceptID'].replace('NCBI:txid', '', inplace=True)
occ.head()

eventID                  eventDate  decimalLatitude  \
0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   
1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   
2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   
3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   
4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   

   decimalLongitude                env_broad_scale  \
0          -121.848  marine biome  (ENVO:00000447)   
1          -121.848  marine biome  (ENVO:00000447)   
2          -121.848  marine biome  (ENVO:00000447)   
3          -121.846  marine biome  (ENVO:00000447)   
4          -121.846  marine biome  (ENVO:00000447)   

                 env_local_scale  \
0  coastal water (ENVO:00001250)   
1  coastal water (ENVO:00001250)   
2  coastal water (ENVO:00001250)   
3  coastal water (ENVO:00001250)   
4  coastal water (ENVO:00001250)   

                                      env_medium target_gene  \
0  waterborne particulate matter (ENVO:01000436)         18S   
1  waterborne particulate matter (ENVO:01000436)         18S   
2  waterborne particulate matter (ENVO:01000436)         18S   
3  waterborne particulate matter (ENVO:01000436)         18S   
4  waterborne particulate matter (ENVO:01000436)         18S   

  pcr_primer_forward        pcr_primer_reverse  ...    kingdom      phylum  \
0   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Eukaryota  Arthropoda   
1   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Eukaryota  Arthropoda   
2   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Eukaryota  Arthropoda   
3   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Eukaryota  Arthropoda   
4   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Eukaryota  Arthropoda   

         class      order         family        genus  \
0  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   
1  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   
2  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   
3  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   
4  Hexanauplia  Calanoida  Paracalanidae  Paracalanus   

                            scientificNameID taxonID nameAccordingTo  \
0  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   
1  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   
2  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   
3  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   
4  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   

  taxonConceptID  
0                 
1                 
2                 
3                 
4                 

[5 rows x 27 columns]

In [97]:
## identificationRemarks

# Get identificationRemarks
occ = occ.merge(plate_meta[['seqID', 'identificationRemarks']], how='left', left_on='eventID', right_on='seqID')
occ.drop(columns='seqID', inplace=True)

# Add name that matched in GenBank
occ['identificationRemarks'] = plate['Species'].copy() + ', ' + occ['identificationRemarks']
occ.head()

eventID                  eventDate  decimalLatitude  \
0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   
1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   
2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   
3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   
4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   

   decimalLongitude                env_broad_scale  \
0          -121.848  marine biome  (ENVO:00000447)   
1          -121.848  marine biome  (ENVO:00000447)   
2          -121.848  marine biome  (ENVO:00000447)   
3          -121.846  marine biome  (ENVO:00000447)   
4          -121.846  marine biome  (ENVO:00000447)   

                 env_local_scale  \
0  coastal water (ENVO:00001250)   
1  coastal water (ENVO:00001250)   
2  coastal water (ENVO:00001250)   
3  coastal water (ENVO:00001250)   
4  coastal water (ENVO:00001250)   

                                      env_medium target_gene  \
0  waterborne particulate matter (ENVO:01000436)         18S   
1  waterborne particulate matter (ENVO:01000436)         18S   
2  waterborne particulate matter (ENVO:01000436)         18S   
3  waterborne particulate matter (ENVO:01000436)         18S   
4  waterborne particulate matter (ENVO:01000436)         18S   

  pcr_primer_forward        pcr_primer_reverse  ...      phylum        class  \
0   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Arthropoda  Hexanauplia   
1   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Arthropoda  Hexanauplia   
2   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Arthropoda  Hexanauplia   
3   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Arthropoda  Hexanauplia   
4   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Arthropoda  Hexanauplia   

       order         family        genus  \
0  Calanoida  Paracalanidae  Paracalanus   
1  Calanoida  Paracalanidae  Paracalanus   
2  Calanoida  Paracalanidae  Paracalanus   
3  Calanoida  Paracalanidae  Paracalanus   
4  Calanoida  Paracalanidae  Paracalanus   

                            scientificNameID taxonID nameAccordingTo  \
0  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   
1  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   
2  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   
3  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   
4  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   

  taxonConceptID                              identificationRemarks  
0                 unassigned, Genbank nr Release 221 September 2...  
1                 unassigned, Genbank nr Release 221 September 2...  
2                 unassigned, Genbank nr Release 221 September 2...  
3                 unassigned, Genbank nr Release 221 September 2...  
4                 unassigned, Genbank nr Release 221 September 2...  

[5 rows x 28 columns]

In [98]:
## basisOfRecord

occ['basisOfRecord'] = 'MaterialSample'
occ.head()

eventID                  eventDate  decimalLatitude  \
0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   
1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   
2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   
3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   
4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   

   decimalLongitude                env_broad_scale  \
0          -121.848  marine biome  (ENVO:00000447)   
1          -121.848  marine biome  (ENVO:00000447)   
2          -121.848  marine biome  (ENVO:00000447)   
3          -121.846  marine biome  (ENVO:00000447)   
4          -121.846  marine biome  (ENVO:00000447)   

                 env_local_scale  \
0  coastal water (ENVO:00001250)   
1  coastal water (ENVO:00001250)   
2  coastal water (ENVO:00001250)   
3  coastal water (ENVO:00001250)   
4  coastal water (ENVO:00001250)   

                                      env_medium target_gene  \
0  waterborne particulate matter (ENVO:01000436)         18S   
1  waterborne particulate matter (ENVO:01000436)         18S   
2  waterborne particulate matter (ENVO:01000436)         18S   
3  waterborne particulate matter (ENVO:01000436)         18S   
4  waterborne particulate matter (ENVO:01000436)         18S   

  pcr_primer_forward        pcr_primer_reverse  ...        class      order  \
0   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Hexanauplia  Calanoida   
1   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Hexanauplia  Calanoida   
2   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Hexanauplia  Calanoida   
3   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Hexanauplia  Calanoida   
4   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Hexanauplia  Calanoida   

          family        genus                           scientificNameID  \
0  Paracalanidae  Paracalanus  urn:lsid:marinespecies.org:taxname:104196   
1  Paracalanidae  Paracalanus  urn:lsid:marinespecies.org:taxname:104196   
2  Paracalanidae  Paracalanus  urn:lsid:marinespecies.org:taxname:104196   
3  Paracalanidae  Paracalanus  urn:lsid:marinespecies.org:taxname:104196   
4  Paracalanidae  Paracalanus  urn:lsid:marinespecies.org:taxname:104196   

  taxonID nameAccordingTo taxonConceptID  \
0  104196           WoRMS                  
1  104196           WoRMS                  
2  104196           WoRMS                  
3  104196           WoRMS                  
4  104196           WoRMS                  

                               identificationRemarks   basisOfRecord  
0  unassigned, Genbank nr Release 221 September 2...  MaterialSample  
1  unassigned, Genbank nr Release 221 September 2...  MaterialSample  
2  unassigned, Genbank nr Release 221 September 2...  MaterialSample  
3  unassigned, Genbank nr Release 221 September 2...  MaterialSample  
4  unassigned, Genbank nr Release 221 September 2...  MaterialSample  

[5 rows x 29 columns]

In [99]:
## Add identificationReferences 

occ = occ.merge(plate_meta[['seqID', 'identificationReferences']], how='left', left_on='eventID', right_on='seqID')
occ.drop(columns='seqID', inplace=True)
occ['identificationReferences'] = occ['identificationReferences'].str.replace('| ', ' | ', regex=False)

occ.head()

eventID                  eventDate  decimalLatitude  \
0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   
1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   
2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   
3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   
4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   

   decimalLongitude                env_broad_scale  \
0          -121.848  marine biome  (ENVO:00000447)   
1          -121.848  marine biome  (ENVO:00000447)   
2          -121.848  marine biome  (ENVO:00000447)   
3          -121.846  marine biome  (ENVO:00000447)   
4          -121.846  marine biome  (ENVO:00000447)   

                 env_local_scale  \
0  coastal water (ENVO:00001250)   
1  coastal water (ENVO:00001250)   
2  coastal water (ENVO:00001250)   
3  coastal water (ENVO:00001250)   
4  coastal water (ENVO:00001250)   

                                      env_medium target_gene  \
0  waterborne particulate matter (ENVO:01000436)         18S   
1  waterborne particulate matter (ENVO:01000436)         18S   
2  waterborne particulate matter (ENVO:01000436)         18S   
3  waterborne particulate matter (ENVO:01000436)         18S   
4  waterborne particulate matter (ENVO:01000436)         18S   

  pcr_primer_forward        pcr_primer_reverse  ...      order         family  \
0   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Calanoida  Paracalanidae   
1   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Calanoida  Paracalanidae   
2   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Calanoida  Paracalanidae   
3   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Calanoida  Paracalanidae   
4   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Calanoida  Paracalanidae   

         genus                           scientificNameID taxonID  \
0  Paracalanus  urn:lsid:marinespecies.org:taxname:104196  104196   
1  Paracalanus  urn:lsid:marinespecies.org:taxname:104196  104196   
2  Paracalanus  urn:lsid:marinespecies.org:taxname:104196  104196   
3  Paracalanus  urn:lsid:marinespecies.org:taxname:104196  104196   
4  Paracalanus  urn:lsid:marinespecies.org:taxname:104196  104196   

  nameAccordingTo taxonConceptID  \
0           WoRMS                  
1           WoRMS                  
2           WoRMS                  
3           WoRMS                  
4           WoRMS                  

                               identificationRemarks   basisOfRecord  \
0  unassigned, Genbank nr Release 221 September 2...  MaterialSample   
1  unassigned, Genbank nr Release 221 September 2...  MaterialSample   
2  unassigned, Genbank nr Release 221 September 2...  MaterialSample   
3  unassigned, Genbank nr Release 221 September 2...  MaterialSample   
4  unassigned, Genbank nr Release 221 September 2...  MaterialSample   

                            identificationReferences  
0  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...  
1  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...  
2  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...  
3  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...  
4  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...  

[5 rows x 30 columns]

In [100]:
## organismQuantity (number of reads)

occ['organismQuantity'] = plate['Reads']
occ['organismQuantityType'] = 'DNA sequence reads'
occ.head()

eventID                  eventDate  decimalLatitude  \
0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   
1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   
2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   
3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   
4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   

   decimalLongitude                env_broad_scale  \
0          -121.848  marine biome  (ENVO:00000447)   
1          -121.848  marine biome  (ENVO:00000447)   
2          -121.848  marine biome  (ENVO:00000447)   
3          -121.846  marine biome  (ENVO:00000447)   
4          -121.846  marine biome  (ENVO:00000447)   

                 env_local_scale  \
0  coastal water (ENVO:00001250)   
1  coastal water (ENVO:00001250)   
2  coastal water (ENVO:00001250)   
3  coastal water (ENVO:00001250)   
4  coastal water (ENVO:00001250)   

                                      env_medium target_gene  \
0  waterborne particulate matter (ENVO:01000436)         18S   
1  waterborne particulate matter (ENVO:01000436)         18S   
2  waterborne particulate matter (ENVO:01000436)         18S   
3  waterborne particulate matter (ENVO:01000436)         18S   
4  waterborne particulate matter (ENVO:01000436)         18S   

  pcr_primer_forward        pcr_primer_reverse  ...        genus  \
0   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Paracalanus   
1   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Paracalanus   
2   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Paracalanus   
3   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Paracalanus   
4   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  Paracalanus   

                            scientificNameID taxonID nameAccordingTo  \
0  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   
1  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   
2  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   
3  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   
4  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   

  taxonConceptID                              identificationRemarks  \
0                 unassigned, Genbank nr Release 221 September 2...   
1                 unassigned, Genbank nr Release 221 September 2...   
2                 unassigned, Genbank nr Release 221 September 2...   
3                 unassigned, Genbank nr Release 221 September 2...   
4                 unassigned, Genbank nr Release 221 September 2...   

    basisOfRecord                           identificationReferences  \
0  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   
1  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   
2  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   
3  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   
4  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   

  organismQuantity organismQuantityType  
0            14825   DNA sequence reads  
1            16094   DNA sequence reads  
2            22459   DNA sequence reads  
3            19312   DNA sequence reads  
4            16491   DNA sequence reads  

[5 rows x 32 columns]

**Note** that there are 215537 rows where the number of reads is 0. These will need to be dropped at the end of the script.

In [101]:
## sampleSizeValue

count_by_seq = plate.groupby('Sequence_ID', as_index=False)['Reads'].sum()
occ = occ.merge(count_by_seq, how='left', left_on='eventID', right_on='Sequence_ID')
occ.drop(columns='Sequence_ID', inplace=True)
occ.rename(columns={'Reads':'sampleSizeValue'}, inplace=True)
print(occ.shape)
occ.head()

(280440, 33)


eventID                  eventDate  decimalLatitude  \
0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   
1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   
2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   
3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   
4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   

   decimalLongitude                env_broad_scale  \
0          -121.848  marine biome  (ENVO:00000447)   
1          -121.848  marine biome  (ENVO:00000447)   
2          -121.848  marine biome  (ENVO:00000447)   
3          -121.846  marine biome  (ENVO:00000447)   
4          -121.846  marine biome  (ENVO:00000447)   

                 env_local_scale  \
0  coastal water (ENVO:00001250)   
1  coastal water (ENVO:00001250)   
2  coastal water (ENVO:00001250)   
3  coastal water (ENVO:00001250)   
4  coastal water (ENVO:00001250)   

                                      env_medium target_gene  \
0  waterborne particulate matter (ENVO:01000436)         18S   
1  waterborne particulate matter (ENVO:01000436)         18S   
2  waterborne particulate matter (ENVO:01000436)         18S   
3  waterborne particulate matter (ENVO:01000436)         18S   
4  waterborne particulate matter (ENVO:01000436)         18S   

  pcr_primer_forward        pcr_primer_reverse  ...  \
0   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   
1   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   
2   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   
3   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   
4   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...   

                            scientificNameID taxonID nameAccordingTo  \
0  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   
1  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   
2  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   
3  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   
4  urn:lsid:marinespecies.org:taxname:104196  104196           WoRMS   

  taxonConceptID                              identificationRemarks  \
0                 unassigned, Genbank nr Release 221 September 2...   
1                 unassigned, Genbank nr Release 221 September 2...   
2                 unassigned, Genbank nr Release 221 September 2...   
3                 unassigned, Genbank nr Release 221 September 2...   
4                 unassigned, Genbank nr Release 221 September 2...   

    basisOfRecord                           identificationReferences  \
0  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   
1  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   
2  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   
3  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   
4  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   

  organismQuantity organismQuantityType sampleSizeValue  
0            14825   DNA sequence reads           85600  
1            16094   DNA sequence reads           90702  
2            22459   DNA sequence reads          130275  
3            19312   DNA sequence reads          147220  
4            16491   DNA sequence reads          121419  

[5 rows x 33 columns]

In [102]:
## sampleSizeUnit

occ['sampleSizeUnit'] = 'DNA sequence reads'
occ.head()

eventID                  eventDate  decimalLatitude  \
0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   
1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   
2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   
3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   
4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   

   decimalLongitude                env_broad_scale  \
0          -121.848  marine biome  (ENVO:00000447)   
1          -121.848  marine biome  (ENVO:00000447)   
2          -121.848  marine biome  (ENVO:00000447)   
3          -121.846  marine biome  (ENVO:00000447)   
4          -121.846  marine biome  (ENVO:00000447)   

                 env_local_scale  \
0  coastal water (ENVO:00001250)   
1  coastal water (ENVO:00001250)   
2  coastal water (ENVO:00001250)   
3  coastal water (ENVO:00001250)   
4  coastal water (ENVO:00001250)   

                                      env_medium target_gene  \
0  waterborne particulate matter (ENVO:01000436)         18S   
1  waterborne particulate matter (ENVO:01000436)         18S   
2  waterborne particulate matter (ENVO:01000436)         18S   
3  waterborne particulate matter (ENVO:01000436)         18S   
4  waterborne particulate matter (ENVO:01000436)         18S   

  pcr_primer_forward        pcr_primer_reverse  ... taxonID nameAccordingTo  \
0   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  104196           WoRMS   
1   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  104196           WoRMS   
2   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  104196           WoRMS   
3   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  104196           WoRMS   
4   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...  104196           WoRMS   

  taxonConceptID                              identificationRemarks  \
0                 unassigned, Genbank nr Release 221 September 2...   
1                 unassigned, Genbank nr Release 221 September 2...   
2                 unassigned, Genbank nr Release 221 September 2...   
3                 unassigned, Genbank nr Release 221 September 2...   
4                 unassigned, Genbank nr Release 221 September 2...   

    basisOfRecord                           identificationReferences  \
0  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   
1  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   
2  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   
3  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   
4  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   

  organismQuantity organismQuantityType sampleSizeValue      sampleSizeUnit  
0            14825   DNA sequence reads           85600  DNA sequence reads  
1            16094   DNA sequence reads           90702  DNA sequence reads  
2            22459   DNA sequence reads          130275  DNA sequence reads  
3            19312   DNA sequence reads          147220  DNA sequence reads  
4            16491   DNA sequence reads          121419  DNA sequence reads  

[5 rows x 34 columns]

In [103]:
## associatedSequences

occ = occ.merge(plate_meta[['seqID', 'associatedSequences']], how='left', left_on='eventID', right_on='seqID')
occ.drop(columns='seqID', inplace=True)
occ.head()

eventID                  eventDate  decimalLatitude  \
0  05114c01_12_edna_1_S  2014-02-20T15:33:00-08:00          36.7958   
1  05114c01_12_edna_2_S  2014-02-20T15:33:00-08:00          36.7958   
2  05114c01_12_edna_3_S  2014-02-20T15:33:00-08:00          36.7958   
3  11216c01_12_edna_1_S  2016-04-21T14:39:00-07:00          36.7962   
4  11216c01_12_edna_2_S  2016-04-21T14:39:00-07:00          36.7962   

   decimalLongitude                env_broad_scale  \
0          -121.848  marine biome  (ENVO:00000447)   
1          -121.848  marine biome  (ENVO:00000447)   
2          -121.848  marine biome  (ENVO:00000447)   
3          -121.846  marine biome  (ENVO:00000447)   
4          -121.846  marine biome  (ENVO:00000447)   

                 env_local_scale  \
0  coastal water (ENVO:00001250)   
1  coastal water (ENVO:00001250)   
2  coastal water (ENVO:00001250)   
3  coastal water (ENVO:00001250)   
4  coastal water (ENVO:00001250)   

                                      env_medium target_gene  \
0  waterborne particulate matter (ENVO:01000436)         18S   
1  waterborne particulate matter (ENVO:01000436)         18S   
2  waterborne particulate matter (ENVO:01000436)         18S   
3  waterborne particulate matter (ENVO:01000436)         18S   
4  waterborne particulate matter (ENVO:01000436)         18S   

  pcr_primer_forward        pcr_primer_reverse  ... nameAccordingTo  \
0   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...           WoRMS   
1   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...           WoRMS   
2   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...           WoRMS   
3   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...           WoRMS   
4   GTACACACCGCCCGTC  TGATCCTTCTGCAGGTTCACCTAC  ...           WoRMS   

  taxonConceptID                              identificationRemarks  \
0                 unassigned, Genbank nr Release 221 September 2...   
1                 unassigned, Genbank nr Release 221 September 2...   
2                 unassigned, Genbank nr Release 221 September 2...   
3                 unassigned, Genbank nr Release 221 September 2...   
4                 unassigned, Genbank nr Release 221 September 2...   

    basisOfRecord                           identificationReferences  \
0  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   
1  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   
2  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   
3  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   
4  MaterialSample  https://github.com/MBARI-BOG/BOG-Banzai-Dada2-...   

  organismQuantity organismQuantityType sampleSizeValue      sampleSizeUnit  \
0            14825   DNA sequence reads           85600  DNA sequence reads   
1            16094   DNA sequence reads           90702  DNA sequence reads   
2            22459   DNA sequence reads          130275  DNA sequence reads   
3            19312   DNA sequence reads          147220  DNA sequence reads   
4            16491   DNA sequence reads          121419  DNA sequence reads   

                            associatedSequences  
0                                           NaN  
1                                           NaN  
2                                           NaN  
3  NCBI BioProject accession number PRJNA433203  
4  NCBI BioProject accession number PRJNA433203  

[5 rows x 35 columns]

In [104]:
## Drop records where organismQuantity = 0 (absences are not meaningful for this data set)

occ = occ[occ['organismQuantity'] > 0]
print(occ.shape)

(64903, 35)


**Note** that the vast majority of this data has 0 reads (215537 records out of 280440, or about 75%). **This makes sense, and is generally fine. Although Katie said something about the 0's being potentially important if the data set were to be reanalyzed due to the compositional nature of the data. We can stay in conversation about this.**

Also **note** that there is one sample (34916c01_12_edna_2_S) that has no reads at all in the ASV table.

In [106]:
## Check for NaN values in string fields

occ.loc[occ['associatedSequences'].isna() == True, 'associatedSequences'] = ''
occ.isna().sum()

eventID                     0
eventDate                   0
decimalLatitude             0
decimalLongitude            0
env_broad_scale             0
env_local_scale             0
env_medium                  0
target_gene                 0
pcr_primer_forward          0
pcr_primer_reverse          0
pcr_primer_name_forward     0
pcr_primer_name_reverse     0
pcr_primer_reference        0
sop                         0
occurrenceID                0
DNA_sequence                0
scientificName              0
kingdom                     0
phylum                      0
class                       0
order                       0
family                      0
genus                       0
scientificNameID            0
taxonID                     0
nameAccordingTo             0
taxonConceptID              0
identificationRemarks       0
basisOfRecord               0
identificationReferences    0
organismQuantity            0
organismQuantityType        0
sampleSizeValue             0
sampleSize

In [107]:
## Divide into occurrence and DNADerivedDataExt

ddd_cols = [
    'eventID',
    'occurrenceID',
    'DNA_sequence',
    'sop',
    'target_gene',
    'pcr_primer_forward',
    'pcr_primer_reverse',
    'pcr_primer_name_forward',
    'pcr_primer_name_reverse',
    'pcr_primer_reference',
    'env_broad_scale',
    'env_local_scale',
    'env_medium',
]

DNADerivedData = occ[ddd_cols].copy()

occ.drop(ddd_cols[2:], axis=1, inplace=True)

## Save

In [108]:
## Save

occ.to_csv('eDNA_occurrence_20210913.csv', index=False, na_rep='NaN')
DNADerivedData.to_csv('eDNA_DNADerivedData_20210913.csv', index=False, na_rep='NaN')

## Questions

1. Katie still needs to remove the clarifying text from the sop column (e.g. 'ext: ') **Done.**